[Spring 체스 - 2단계] 포키(권예진) 미션 제출합니다.

YJGwon opened the pull request:
안녕하세요 또링!!
1단계에서 말씀주신 Controller Test는 아직 작성하지 못했어요..죄송합니다 😢 
이것까지 하면 리뷰요청이 너무 늦어질 것 같아서.. 일단 구현 완료하고 요청 드립니다.

---

#### 요구사항 변경에 따른 도메인 수정
2단계 요구사항에 맞춰 여러 방에서 진행할 수 있도록 하려다 보니 **Domain 수정이 많이 일어났어요**.
- Camp에서 turn을 관리하고 있어 Application 전역적으로 turn정보가 통일되던 점
- 상태 패턴으로 인해 종료버튼으로 Game을 종료하지 않으면 다음 동작이 불가능했던 점

크게 두 가지 문제가 수정되었습니다.


#### Http Method 변경
HttpMethod는 나름대로 학습한 후 적절하게 바꿔보았어요. 이 과정에서 PUT method 사용을 위한 spring 설정이 추가되었습니다.

#### Error 처리
Error 처리에 관해서도 고민해보았는데요,
같은 IllegalArgumentException이어도 **상황에 따라 status가 달라져야 해서 결국 try-catch를 사용**하게 되었습니닷...
그러나 **javascript에서 이해할 수 있게 메세지를 보내는 작업이 혹시 view의 영향을 받고있는건가?** 하는 우려가 있어요. 이 부분에 대한 또링의 코멘트가 궁금하(..다기보다 두렵..)네요!


---

1단계에 비해 많이 자신없는 상태여서 스스로도 아쉽고 무엇보다 리뷰해주실 또링에게 죄송하네요ㅠㅠ 절대 주말을 넘기고 싶지 않아서 요청 드리는 제 마음을 너그러이 이해해주시면 감사하겠습니다.. 😭

요즘 일교차가 심한데 건강 잘 챙기셨으면 좋겠어요!! 이번 리뷰도 (염치없지만) 잘 부탁드립니다!

YJGwon committed:
chore(web/GameController): 필요없는 main method 삭제

YJGwon committed:
docs(README.md): 2단계 구현목록 작성

YJGwon committed:
feat(resources/templates): 방 입장 페이지 추가, 게임 진행 페이지 분리

YJGwon committed:
feat: index 페이지 요청 시 DB에 저장된 게임방 목록 조회

- game table 수정: 필요한 column 추가
- game table의 column 구조 바뀌면서 game관련 sql문 임시로 수정
- GameDto 추가: db의 게임 데이터 담아오기 위해

YJGwon committed:
feat: 제목과 비밀번호로 방 생성하여 DB 저장

YJGwon committed:
feat: 방 생성 후 방 목록 새로고침

YJGwon committed:
feat: 방 제목 클릭 시 해당하는 체스판 load

YJGwon committed:
feat: 여러 게임을 동시에 진행할 수 있도록 상태패턴 제거

- Application 전역에서 공통으로 관리되던 turn 정보 게임 객체 안으로 이동
- 매번 요청 수신할 때 마다 DB에서 게임 로드한 뒤 처리하도록 수정
- 게임 진행 상황을 매번 자동으로 DB에 저장하도록 수정
- console application 삭제

YJGwon committed:
feat: 게임방 입장 시 DB에 저장된 비밀번호 확인

YJGwon committed:
feat: 게임 종료 상태 DB에 저장

YJGwon committed:
feat: 비밀번호, 진행 상태 확인 후 게임방 삭제

YJGwon committed:
refactor: API 설계 수정

- PUT method 사용 위한 hidden method filter 설정 추가
- url에서 동사 대신 리소스명 사용

YJGwon committed:
refactor: 에러 코드, 메세지 수정

YJGwon committed:
chore: 입장 시 비밀번호 입력 삭제

- 요구사항에 따르면 입장 시에는 비밀번호 확인 불필요

kang-hyungu commented on src/main/java/chess/application/web/GameController.java:
```diff
@@ -20,60 +26,91 @@ public GameController(GameService gameService) {
-        return play(model);
+    @GetMapping("/games")
+    public ResponseEntity<String> list() {
+        String listData = jsonTransformer.render(gameService.list());
```
1. RestController를 쓰면 jsonTransformer가 필요 없을 것 같네요.
2. 변수명에 list, data 같은 추상적인 표현 대신 명확한 이름을 붙여주세요. 리스트데이터가 체스게임의 무엇을 말하는 걸까요?

kang-hyungu commented on src/main/java/chess/application/web/GameController.java:
```diff
@@ -20,60 +26,91 @@ public GameController(GameService gameService) {
+        try {
+            gameService.createRoom(title, password);
+        } catch (IllegalArgumentException e) {
+            return createMessageResponse(HttpStatus.BAD_REQUEST, e.getMessage());
```
컨트롤러마다 createMessageResponse를 만들면 중복 코드가 많아질 것 같네요.
스프링에서 제공하는 ControllerAdvice와 ExceptionHandler 키워드를 검색해보고 예외를 한 곳에서 처리하도록 만들어보아요

YJGwon replied:
ControllerAdvice를 적용한 ExceptionController를 만들고, exception 발생 시 메세지와 함께 error page로 redirect하도록 수정했습니다.

kang-hyungu commented on src/main/java/chess/application/web/GameController.java:
```diff
@@ -20,60 +26,91 @@ public GameController(GameService gameService) {
+    public ResponseEntity<String> delete(@PathVariable long gameNo, @RequestBody String password) {
+        try {
+            gameService.checkPassword(gameNo, password);
+            gameService.delete(gameNo);
```
컨트롤러에서 여러 메서드를 호출하기보다는 서비스의 메서드 하나에서 다 처리하도록 만들어주세요.

```suggestion
    try {
            gameService.delete(gameNo);
            return ResponseEntity.ok("방이 삭제되었습니다.");
    } catch (IllegalArgumentException e) {
```

kang-hyungu commented on src/main/java/chess/application/web/GameController.java:
```diff
@@ -20,60 +26,91 @@ public GameController(GameService gameService) {
+                       @RequestParam String source, @RequestParam String target) {
+        ChessGame chessGame = gameService.load(gameNo);
+        gameService.move(source, target, chessGame);
+        gameService.save(gameNo, chessGame);
```
여기도 마찬가지로 서비스의 메서드 하나만 호출하도록 만들어보세요.

kang-hyungu commented on src/main/java/chess/application/web/GameController.java:
```diff
@@ -20,60 +26,91 @@ public GameController(GameService gameService) {
-        model.addAllAttributes(gameService.end());
+    private String play(Model model, long gameNo) {
+        ChessGame chessGame = gameService.load(gameNo);
+        if (gameService.isGameFinished(gameNo, chessGame)) {
```
컨트롤러에 게임 로직과 관련된 if절이 생기는 건 적절하지 않습니다.
서비스 안으로 옮겨보세요.
컨트롤러 내 다른 메서드도 같이 점검해보아요.

kang-hyungu commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -7,76 +7,74 @@
-    public void start() {
-        chessGame.start();
+    public void createRoom(String title, String password) {
+        if (title.isEmpty() || password.isEmpty()) {
```
1. isBlank로 바꿔봅시다.
2. isEmpty와 isBlank의 차이는 무엇일까요?

YJGwon replied:
isEmpty는 length만 검사하고, isBlank는 공백만 있는 경우도 검사합니다. 변경했습니다!

kang-hyungu commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -7,76 +7,74 @@
+        if (title.isEmpty() || password.isEmpty()) {
+            throw new IllegalArgumentException("방 제목과 비밀번호를 입력하세요.");
+        }
+        long gameNo = gameDao.insert(GameDto.fromNewGame(title, password));
```
dao 관련 처리에 dto가 필요할까요?
dto를 제거 할 방법 없을까요?

kang-hyungu commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -7,76 +7,74 @@
+            throw new IllegalArgumentException("방 제목과 비밀번호를 입력하세요.");
+        }
+        long gameNo = gameDao.insert(GameDto.fromNewGame(title, password));
+        boardDao.insert(gameNo, BoardInitializer.get().getSquares());
```
객체지향 생활체조
규칙 4: 한 줄에 점을 하나만 찍는다.

kang-hyungu commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -7,76 +7,74 @@
-        model.put(KEY_READY, false);
-        return model;
+    public void checkPassword(long gameNo, String password) {
+        if (!gameDao.loadPassword(gameNo).equals(password)) {
```
game 테이블에 필드가 하나씩 추가 될때마다 dao도 영향을 받겠군요.
game 객체를 따로 만들지 않은 이유가 있을까요?

kang-hyungu commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -7,76 +7,74 @@
+        List<PieceDto> rawBoard = boardDao.load(gameNo);
         Map<Position, Piece> board = rawBoard.stream()
                 .collect(Collectors.toMap(
                         pieceDto2 -> parsePosition(pieceDto2.getPosition()),
```
pieceDto2 이름을 바꿔봅시다


---

클린코드 25p

의미 있게 구분하라
컴파일러나 인터프리터만 통과하려는 생각으로 코드를 구현하는 프로그래머는 스스로 문제를 일으킨다.
예를 들어, 동일한 범위 안에서는 다른 두 개념에 같은 이름을 사용하지 못한다.
그래서 프로그래머가 한쪽 이름을 마음대로 바꾸고픈 유혹에 빠진다.
어떤 프로그래머는 철자를 살짝 바꿨다가 나중에 철자 오류를 고치는 순간 컴파일이 불가능한 상황에 빠진다.(예를 들면 변수명을 klass로 짓는 경우)
컴파일러를 통과할지라도 연속된 숫자를 덧붙이거나 불용어(noise word)를 추가하는 방식은 적절하지 못하다.
이름이 달라야 한다면 의미도 달라져야 한다.
연속적인 숫자를 덧붙인 이름(a1, a2, ... aN)은 의도적인 이름과 정반대다.
이런 이름은 그릇된 정보를 제공하는 이름도 아니며, 아무런 정보를 제공하지 못하는 이름일 뿐이다.
저자 의도가 전혀 드러나지 않는다.

YJGwon replied:
PieceDto를 수정하면서 PieceDto2를 만들어서 점진적인 refactoring을 시도했는데, 바꾸지 않은 변수명이 남아있었네요 ㅠㅠ 수정했습니다!

kang-hyungu commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -108,43 +96,44 @@ private Position parsePosition(String rawPosition) {
-    public Map<String, Object> modelStatus() {
-        Map<Camp, Score> scores = chessGame.getScores();
-        return scores.entrySet().stream()
+    public Map<String, Object> modelStatus(ChessGame chessGame) {
```
서비스에서 처리할 필요가 없는 메서드는 적절한 도메인 객체에서 처리하도록 합니다.
메서드 중에 dao를 쓰지 않는 메서드는 모두 도메인 객체로 옮겨보세요

kang-hyungu reviewed (changes requested):
포키 2단계 구현 잘 해주셨네요.
몇 가지 피드백 남겼는데 적용해보시고 리뷰 재요청주세요.
궁금한 점 있으면 오프라인으로 찾아오셔도 좋습니다.

YJGwon committed:
refactor: move, delete 한 service method로 처리

YJGwon committed:
fix(GameService): 제목과 비밀번호에 공백만 입력된 경우에도 예외 발생

YJGwon committed:
refactor(GameService): PieceDto refactor 후 수정되지 않았던 변수명 수정

YJGwon committed:
refactor(ExceptionController): ExceptionController에서 error page로 redirect

YJGwon committed:
refactor(application/web): RestController 추가

- JsonTransformer 삭제
- GameService 내 load() 호출 후 실행되어야 하는 method들은 내부에서 load() 호출하게 수정

YJGwon committed:
refactor(domain/Room): 게임 방을 나타내는 Room 추가

- 게임 방과 게임 객체 분리, DB Table Room 추가
- GameDto 삭제

YJGwon committed:
refactor: Room에 field로 running 추가, Service에서 도메인 로직 분리

- 체스판 상에서 판단되는 게임 종료와 사용자가 임의로 종료한 상황 분리
- Room 필드에 맞게 schema 수정: RoomDao에서 Room 객체를 만들 수 있도록
- Service에서 room table field를 각각 조회하지 않고 Room 객체 load하여 처리하도록 수정
- 불필요한 console print 구문 삭제

YJGwon commented:
안녕하세요 구구! 포키입니다. 리뷰 요청이 많이 늦어졌네요..
최대한 체스 미션 내에서 할 수 있는 고민들을 다 해보고 끝내고 싶었는데 여전히 DB, Service와 Domain이 깔끔히 분리되진 않은 듯 합니다.
좋은 comment 많이 주셨는데 다 반영하지 못한 것에 대해 🍎드립니다 😭 

- 🍎 1: `modelSataus()` 쪽에서 dao를 사용하지 않는 로직을 제거하라고 comment 주셨는데 view에서 알아들을 수 있는 형태로 modeling하는 것이 domain의 역할은 아닌 것 같아 일단은 그대로 두었어요.
- 🍎 2: refactoring 하다보니 commit 단위가 너무 커졌습니다 ㅠㅠ 고려해본 적 없는 것들에 대해 고민하다보니까 작은 단위로 개발하는 것이 쉽지 않더라구요...
- 🍎 3: error에 대한 view 처리를 좀 naive하게 했습니다(양심고백)
- 🍎 4: 구구는 다음 미션때 다시 고민해도 좋다 하셨지만 Controller의 if절을 없애지 못한 것 역시 좀 아쉽긴 합니다

어쩌다보니까 PR이 아니고 반성문(같은 변명문)이 되어버렸는데요..! 이러다간 지하철 미션 끝날때까지 체스 리팩토링 하고 있게 될 것 같아서 스스로 PR제출 기한을 수요일로 정했어요. 대신 그 안에 새로 흡수한 관점들을 조금이라도 코드를 통해 표현해보려고 시도했습니다. 

우테코 이끌어주시느라 바쁘신 와중에 리뷰도 꼼꼼히 해주시고 좋은 말씀 많이 해주셔서 감사합니다!
궁금한게 있을 때 또 찾아뵙겠습니다 😄

kang-hyungu commented on src/main/java/chess/application/web/ExceptionController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
```
bad request는 클라이언트 에러인데 모든 exception을 4xx 대 에러로 보긴 어려울 것 같네요.

YJGwon replied:
Internal Server Error로 변경했습니다! Spring을 사용하게 되니까 확실히 예외 type을 신중하게 골라야겠다는 생각이 드네요 😅

kang-hyungu commented on src/main/java/chess/application/web/GameRestController.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @GetMapping("/score/{gameNo}")
+    public ResponseEntity<Map<String, Object>> status(@PathVariable int gameNo) {
```
응답을 map 대신 dto로 감싸주면 나중에 코드 관리하기가 더 좋겠네요

kang-hyungu commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -125,15 +98,39 @@ private Map<String, Object> modelResult(ChessGame chessGame) {
+        roomDao.delete(roomNo);
+    }
+
+    private void checkPassword(Room room, String password) {
```
이 메서드는 Room 클래스에 추가할 수 있겠네요

kang-hyungu commented on src/main/java/chess/application/web/GameService.java:
```diff
@@ -125,15 +98,39 @@ private Map<String, Object> modelResult(ChessGame chessGame) {
 
-    public void delete(long gameNo) {
-        if (gameDao.isRunning(gameNo)) {
+    private void checkStatus(Room room) {
```
이 메서드도 Room 클래스에 추가할 수 있겠네요.
예외 처리도 일종의 게임 규칙으로 볼 수 있을테니깐요
Room에 delete 라는 메서드가 있었으면 좀 더 명확하게 보였을거에요.

YJGwon replied:
지하철 미션도 끝난 지금이지만 여전히 Web에서 Domain 로직을 응집시키는 것이 쉽지 않네요..! 꾸준히 고민하고 더 신중하게 구현해보겠습니다 😄

kang-hyungu reviewed (approved):
포키 피드백 반영 잘 하셨어요.
컨트롤러의 if문 제거가 쉽지 않았을텐데 잘 분리해주셨네요
pr은 머지 처리하겠습니다. 수고하셨어요.

YJGwon commented:
늦었지만 코멘트 주신 부분들 간단하게나마 제 repo에 반영해보았습니다! 지하철 하느라 리뷰를 너무 늦게 반영해서 죄송하고, 마지막까지 좋은 코멘트 많이 남겨주셔서 감사해요.
